### PR TITLE
DAOS-9932 control: Add tmpfs scm support to config generate

### PR DESCRIPTION
**Pull request by tanabarr**
Add support for using tmpfs/ramdisk instead of PMem, if
--use-tmpfs-scm is set when calling config generate command, add
ram-disks in the server config and calculate size to use (in total
across all engines) 75% of totalmemory (as reported by
/proc/meminfo MemTotal metric).

Change --min-nr-ssds (int) option to --scm-only (bool) to simplify
usage. Rather then specifying zero for minimum number of SSDs set flag
to disable NVMe otherwise attempt to use all SSDs.

Update the admin guide to reflect the following changes in config
generate commands.
- Add daos_server config generate command (in addition to dmg variant).
- Support --use-tmpfs-scm to generate RAM-disk SCM configs.
- Change --min-nr-ssds to --scm-only to improve usability.

Features: control

Required-githooks: true
Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by tanabarr** [comment]
Existing CI failures (as currently expected when running with `Features: control` pragma):

https://daosio.atlassian.net/browse/DAOS-12013
https://daosio.atlassian.net/browse/DAOS-12326

**Review by knard-intel — APPROVED** [review]
LGTM for what I understand.

The failing test Test Hardware / Functional Hardware Medium / 1-./container/snapshot_aggregation.py seems to have been solved since the PR was pushed https://daosio.atlassian.net/browse/DAOS-12326 and thus should not be related to this PR

The failing test Functional Hardware Medium / 1-./control/dmg_telemetry_io_latency.py seems to be under review https://daosio.atlassian.net/browse/DAOS-12013 and thus should not be related to this PR

File: src/tests/ftest/util/dmg_utils.py [hunk 1]
```diff
@@ -1121,16 +1121,16 @@ def pool_evict(self, pool):
...
                         net_class=None, net_provider=None):
         """Produce a server configuration.
 
         Args:
             access_points (str): Comma separated list of access point addresses.
             num_pmem (int): Number of SCM (pmem) devices required per
                 storage host in DAOS system. Defaults to None.
-            num_nvme (int): Minimum number of NVMe devices required per storage
-                host in DAOS system. Defaults to None.
+            scm_only (bool, option): Whether to omit NVMe from generated config.
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/tests/ftest/util/dmg_utils.py:1132]: done

File: src/tests/ftest/util/dmg_utils.py [hunk 2]
```diff
@@ -1121,20 +1121,22 @@ def pool_evict(self, pool):
...
+                Defaults to False.
             net_class (str): Network class preferred. Defaults to None.
                 i.e. "ethernet"|"infiniband"
             net_provider (str): Network provider preferred. Defaults to None.
                 i.e. "ofi+tcp;ofi_rxm"|"ofi+psm2" etc.
+            use_tmpfs_scm (bool, option): Whether to use a ramdisk instead of PMem
+                as SCM. Defaults to False.
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/tests/ftest/util/dmg_utils.py:1139]: done

File: src/tests/ftest/control/config_generate_output.py [hunk 3]
```diff
@@ -230,6 +230,80 @@ def test_basic_config(self):
...
+        :avocado: tags=ConfigGenerateOutput,test_tmpfs_scm_config
+        """
+        # Get necessary storage and network info.
+        self.prepare_expected_data()
+
+        # Call dmg config generate.
+        result = self.get_dmg_command().config_generate(
+            access_points="wolf-a", net_provider=self.def_provider, use_tmpfs_scm=True)
+        if result.exit_status != 0:
+            errors.append("Config generate failed with use_tmpfs_scm = True!")
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/tests/ftest/control/config_generate_output.py:261]: done

**Comment by mjmac** [comment]
I will do an in-depth review of the code later, but I did check out the PR to play around with it this morning.

The PR as-is does work, but it's not as user friendly as I think it could be.

On my wolf node, I have no pmem and one infiniband adapter. There is an SSD, but it's on the "wrong" NUMA node:
```
NUMA Node 0
  CPU cores: 0-21
  PCI buses:
    0000:[00-07]
      0000:00:11.4
        0000:00:11.4 sdd (1.0 TB block device)
        0000:00:11.4 sdb (1.0 TB block device)
        0000:00:11.4 sdc (1.0 TB block device)
        0000:00:11.4 sda (1.0 TB block device)
      0000:00:1f.2
        0000:00:1f.2 sdf (1.0 TB block device)
        0000:00:1f.2 sdg (200 GB block device)
        0000:00:1f.2 sde (1.0 TB block device)
      0000:02:00.0
        0000:02:00.0 ib0 (network interface) @ 7.88 GB/s
        0000:02:00.0 mlx4_0 (OFI domain) @ 7.88 GB/s
      0000:03:00.0
        0000:03:00.0 eth0 (network interface) @ 4.92 GB/s
      0000:03:00.1
        0000:03:00.1 eth1 (network interface) @ 4.92 GB/s
    0000:7f
    0000:ff
NUMA Node 1
  CPU cores: 22-43
  PCI buses:
    0000:[80-81]
      0000:81:00.0
        0000:81:00.0 nvme0n1 (400 GB block device) @ 3.94 GB/s
Virtual Devices
  lo (network interface)
```

Obviously not an ideal server to run DAOS on, but it should still be possible. I can force the config generator to make me a config given the correct incantations (using ethernet because ib is down):
```
$ daos_server config generate -s -c ethernet -t -e 1
NOTICE: Configuration includes only one access point. This provides no redundancy in the event of an access point failure.
port: 10001
transport_config:
  allow_insecure: false
  client_cert_dir: /etc/daos/certs/clients
  ca_cert: /etc/daos/certs/daosCA.crt
  cert: /etc/daos/certs/server.crt
  key: /etc/daos/certs/server.key
engines:
- targets: 16
  nr_xs_helpers: 0
  first_core: 0
  log_file: /tmp/daos_engine.0.log
  storage:
  - class: ram
    scm_mount: /mnt/daos0
    scm_size: 26
  provider: ofi+tcp;ofi_rxm
  fabric_iface: eth0
  fabric_iface_port: 31416
  pinned_numa_node: 0
disable_vfio: false
disable_vmd: false
enable_hotplug: false
nr_hugepages: 8192
disable_hugepages: false
control_log_mask: INFO
control_log_file: /tmp/daos_server.log
core_dump_filter: 19
name: daos_server
socket_dir: /var/run/daos_server
provider: ofi+tcp;ofi_rxm
access_points:
- localhost:10001
fault_cb: ""
hyperthreads: false
```

It would be nice if the generator just did its best to make a config based on what it finds, even if it's not ideal. I'm fine with doing that as a followup, though. One thing that I did notice while playing around is that it seems to be looking at the memory available rather than the total memory when deciding how big to make the tmpfs. I don't think this is correct because we want the generator to be deterministic given specific hardware configurations.

**Comment by tanabarr** [comment]
https://github.com/daos-stack/daos/pull/11213 is the follow-up documentation PR

@mjmac in terms of immediate changes to this PR, would you like to use a specific percentage of `mem_total` as opposed to `mem_available` as reported by `/proc/meminfo`? I'm fine with making this change, is just not what was decided upon when setting requirements with the team initially. If so, what percentage should we start with? 75% as it is now?

**Comment by tanabarr** [comment]
as requested now using `mem_total` Rather than `mem_available` to calculate tmpfs/ramdisk size

**Review by phender — APPROVED** [review]

**Comment by tanabarr** [comment]
Existing CI failures (as currently expected when running with Features: control pragma):

      https://daosio.atlassian.net/browse/DAOS-12013

File: src/tests/ftest/control/config_generate_output.py [hunk 4]
```diff
@@ -352,64 +426,59 @@ def test_num_engines(self):
...
-            if result.exit_status != 0:
-                errors.append("config generate failed with min_ssd = {}!".format(num_ssd))
-
-        # Call dmg config generate --min_ssds=<min_ssd + 1>. Should fail.
+        # Call dmg config generate with --scm-only=False
         result = dmg.config_generate(
-            access_points="wolf-a", min_ssds=min_ssd + 1, net_provider=self.def_provider)
-        if result.exit_status == 0:
-            errors.append("config generate succeeded with min_ssd + 1 = {}!".format(min_ssd + 1))
+            access_points="wolf-a", scm_only=False, net_provider=self.def_provider)
```
Comment (knard-intel) [inline src/tests/ftest/control/config_generate_output.py:452]: Not sure to fully understand test test suite. However, is this test limited to the wolf test cluster A as the acces_points parameters is hard coded?

**Review by knard-intel — COMMENTED** [review]
Mostly OK for me, there is just some points with ftest on which I am not sure of the consequences.

File: src/control/lib/control/auto.go [hunk 5]
```diff
@@ -878,9 +902,81 @@ func correctSSDCounts(log logging.Logger, minNrSSDs int, sd *storageDetails) err
...
+		if err != nil {
+			return nil, errors.Wrapf(err, "calculate scm ram size")
+		}
+
+		// convert from bytes to gib for server config ramdisk size param
+		sizeGiB := uint(size / humanize.GiByte)
+		if sizeGiB < ramdiskMinSizeGiB {
+			log.Errorf("available memory for scm ramdisk too small, want %d GiB have "+
+				"%s (%d bytes)", ramdiskMinSizeGiB, humanize.IBytes(uint64(size)),
+				size)
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/control/auto.go:937]: The minimum size value is not based on any concrete limitation and it's just a recommendation, I don't want to preclude edge cases where a small size is required for a specific use. Maybe in a future iteration we can come up with a hard limit that will return an error.

File: src/control/lib/control/auto.go [hunk 6]
```diff
@@ -878,9 +902,81 @@ func correctSSDCounts(log logging.Logger, minNrSSDs int, sd *storageDetails) err
 	return nil
 }
 
+// Calculate RAM-disk size (in bytes) based on available memory as reported by /proc/meminfo and the
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/control/auto.go:905]: I will correct this in a follow-up or if I re-push.

File: src/control/lib/control/auto.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (knard-intel) [reply src/control/lib/control/auto.go:937]: OK, thanks for the explanation.

File: src/tests/ftest/control/config_generate_output.py (hunk 4 above)
Comment (knard-intel) [reply src/tests/ftest/control/config_generate_output.py:452]: OK, thanks for the explanation.

**Review by knard-intel — APPROVED** [review]
OK for me.

**Review by mjmac — APPROVED** [review]

**Pull request merged**